### ISSUE TITLE
make literalinclude refer to snippets

### ISSUE THREAD

**Issue by tiagomacarios** [sphinx-doc/sphinx#5597]
Hi would it be possible to make literalinclude refer to snippets in a file?

I want to document C++ code and I want to make sure that all the code being documented actually compiles. So I would rather not have code blocks, but instead "include" code from a file that is being compiled as part of the build.

Sphinx already has the ability to refer to files and to file snippets thorough literalinclude + lines, but it would be great if I could just literalinclude the file and refer to a tag. This way I can modify the source file and not worry about Sphinx missing/adding lines. Sphinx would read the file and only display the lines inside the refered tag.

The QT documentation does this and it is an awesome feature. Here is an example of the [finished product](http://doc.qt.io/qt-5/qtwidgets-mainwindows-application-example.html), the [raw documentation file](https://github.com/qt/qtbase/blob/c5307203f5c0b0e588cc93e70764c090dd4c2ce0/examples/widgets/doc/src/application.qdoc), and the [source code](https://github.com/qt/qtbase/blob/c5307203f5c0b0e588cc93e70764c090dd4c2ce0/examples/widgets/mainwindows/application/mainwindow.cpp).

Basically the source code has tags [number], example:
```
//! [0]
#include <QtWidgets>

#include "mainwindow.h"
//! [0]
```

The docs refer to those tags:

`\snippet mainwindows/application/mainwindow.h 0`

**Comment by tk0miya** [comment]
How about this?
```
.. literalinclude:: mainwindow.cpp
   :start-after: //! [0]
   :end-before: //! [0]
```

This try to filter a file content from line contains `//! [0]` to line contains `//![0]`.

**Comment by tiagomacarios** [comment]
I wasn't aware of that. Let me try it and I'll report back.

**Comment by tk0miya** [comment]
I'm closing this now because this is not an issue of Sphinx.
Thanks,
